### PR TITLE
The port of NanoUI itself

### PR DESCRIPTION
**Pull request by DominikPanic**
Won't work without the previous PR.
/**

    jQuery.timers - Timer abstractions for jQuery
    Written by Blair Mitchelmore (blair DOT mitchelmore AT gmail DOT com)
    Licensed under the WTFPL (http://sam.zoy.org/wtfpl/).
    Date: 2009/10/16
    @author Blair Mitchelmore
    @Version 1.2

**/
Before: https://puu.sh/DkLpr/672ccb17d5.png
After: https://puu.sh/DlFle/d448e50a25.png
(both done on the server with the same amount of people)
(note the JSON procs being replaced)

### CODE REVIEW

File: code/modules/nano/nanoui.dm [hunk 1]
```diff
@@ -438,22 +472,27 @@ nanoui is used to open and update nano browser uis
 		return
 	var/params = "\ref[src]"
 
-	winset(user, window_id, "on-close=\"nanoclose [params]\"")
+	spawn(2)
```
Comment (Rohesie) [inline code/modules/nano/nanoui.dm]: CALLBACK here.

File: code/modules/nano/nanoui.dm [hunk 2]
```diff
@@ -438,22 +472,27 @@ nanoui is used to open and update nano browser uis
...
+		if(!user || !user.client)
```
Comment (Rohesie) [inline code/modules/nano/nanoui.dm]:
```suggestion
		if(!user?.client)
```

File: code/controllers/stonedmc/subsystem/nano.dm [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  * This is called when a player transfers from one mob to another
+  * Transfers all open UIs to the new mob
+  *
+  * @param oldMob /mob The user's old mob
+  * @param newMob /mob The user's new mob
+  *
+  * @return nothing
+  */
+/datum/controller/subsystem/processing/nano/proc/user_transferred(mob/oldMob, mob/newMob)
+	if(!oldMob || !oldMob.open_uis)
```
Comment (Rohesie) [inline code/controllers/stonedmc/subsystem/nano.dm]:
```suggestion
	if(!oldMob?.open_uis)
```

File: code/modules/client/asset_cache.dm [hunk 4]
```diff
@@ -534,4 +534,43 @@ GLOBAL_LIST_EMPTY(asset_datums)
...
+		"nano/templates/"
+	)
+
+
+/datum/asset/nanoui/register()
+	// Crawl the directories to find files.
+	for(var/path in common_dirs)
+		var/list/filenames = flist(path)
+		for(var/filename in filenames)
+			if(copytext(filename, length(filename)) != "/") // Ignore directories.
```
Comment (Rohesie) [inline code/modules/client/asset_cache.dm]: Early continue.

File: code/modules/client/asset_cache.dm [hunk 5]
```diff
@@ -534,4 +534,43 @@ GLOBAL_LIST_EMPTY(asset_datums)
...
+				if(fexists(path + filename))
+					common[filename] = fcopy_rsc(path + filename)
+					register_asset(filename, common[filename])
+	
+	for(var/path in uncommon_dirs)
+		var/list/filenames = flist(path)
+		for(var/filename in filenames)
+			if(copytext(filename, length(filename)) != "/") // Ignore directories.
```
Comment (Rohesie) [inline code/modules/client/asset_cache.dm]: Early continue.

File: code/modules/nano/nanoexternal.dm [hunk 6]
```diff
@@ -7,23 +7,23 @@
...
-			src.Topic(href, params2list(href), ui.ref)	// this will direct to the atom's Topic() proc via client.Topic()
-		else if (ui.on_close_logic)
+			Topic(href, params2list(href), ui.ref)	// this will direct to the atom's Topic() proc via client.Topic()
+		else if(ui.on_close_logic)
 			// no atomref specified (or not found)
 			// so just reset the user mob's machine var
-			if(src && src.mob)
-				src.mob.unset_interaction()
+			if(mob)
+				mob.unset_interaction()
```
Comment (Rohesie) [inline code/modules/nano/nanoexternal.dm]:
```suggestion
				mob?.unset_interaction()
```
And remove the `if` check.

**Comment by Rohesie** [comment]
Still WIP?

**Comment by DominikPanic** [comment]
Yes this will be WIP until you merge the other PR

**Comment by DominikPanic** [comment]
And done

**Pull request merged**